### PR TITLE
Reschedule validations

### PR DESCRIPTION
**Pull request by neezer**
Do them on an explicit call rather than initialization. Mainly for use cases where we want to use this library with "less complete" JWT payloads (such as with testing).

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -83,6 +68,18 @@ class JwtClient {
       tenants: this._payload.tenants
     }
   }
+
+  validate (propsToValidate = []) {
+    const dump = this.dump()
+
+    propsToValidate.forEach(prop => {
+      if (dump[prop] === null || dump[prop] === undefined) {
```
Comment (blobor) [inline index.js]: I believe you told that we are validating shape of payload, not the value of it. Object could have an `id` prop, but it's value could be `null`.

File: index.js [hunk 2]
```diff
@@ -6,26 +6,11 @@ class JwtClient {
   constructor (jwt) {
     this._jwt = this._split(jwt)
     this._payload = this._decode(this._jwt)
-
-    this._assertShape(this._payload)
-
     this._userId = this._payload.id
```
Comment (blobor) [inline index.js]: In this case, if payload is `null` (invalid jwt), it will throw an error.

File: index.js (hunk 1 above)
Comment (neezer) [reply index.js]: Yeah, but `null` is a singleton value, so it's always wrong if it is null: it's antithetical to the meaning of the property.

I suppose technically you could just check for `undefined`, but I'm struggling to think of a situation in which a `null` value is acceptable--if you're checking for the property.

File: index.js (hunk 2 above)
Comment (neezer) [reply index.js]: Yup, still does.

File: index.js (hunk 1 above)
Comment (neezer) [reply index.js]: > so it's always wrong if it is null

To clarify, `id` might be a

- UUID
- random string
- number
- float
- lyrics to a Bob Dylan song

... depending on business rules. But `id` can never be null accoriding to business rules... unless our business is run by the Borg:

ec8b47d56535c9d7b4cc3aa068943e7b001f3fbd0f07e5449db27f76381c4809

File: index.js (hunk 2 above)
Comment (blobor) [reply index.js]: In this case what purpose of custom errors?

File: index.js [hunk 3]
```diff
@@ -83,6 +68,18 @@ class JwtClient {
       tenants: this._payload.tenants
     }
   }
+
+  validate (propsToValidate = []) {
```
Comment (blobor) [inline index.js]: I believe we make this package for abstracting out jwt structure and validation, but looks like now validation is responsibility of package user.

Comment (blobor) [reply index.js]: More over, user should be aware of `jwt` payload structure to pass correct props for validation.

Comment (neezer) [reply index.js]: > user should be aware of jwt payload structure to pass correct props for validation.

No, they need to know the public API, not the internal structure. We do the check off of `dump()`, not off of `this._payload`.

But this does mean my docs update is wrong: `id` should be `userId`.

Comment (blobor) [reply index.js]: In this case we have one more confusion, cause `dump()` returns object with `id` prop which is equal to `jwtClient(jwt).userId`.

Comment (neezer) [reply index.js]: @blobor https://github.com/kofile/jwt-client/pull/12

File: index.js (hunk 2 above)
Comment (neezer) [reply index.js]: My take on [@VitaliiNobis 's issue](https://github.com/kofile/jwt-client/issues/1)

> But current errors are generic and cannot be differentiated from other runtime errors.

Was that if you had a block like this

```js
try {
  const client = makeJwt(jwt)
  const payload = { id: client.userId }

  await userService.post('/users', payload)
} catch (error) {
  // how to determine if the error was the fault of `makeJwt(jwt)`
  // or `userService.post()`??
}
```

... you could be more intelligent about handling that error. Both errors thrown at instantiation (in the case where `jwt === undefined` above) and at validation can be "compared" using `error.name === 'JwtError'`.

If I misunderstood that request, I'd love some clarification.

**Pull request merged**